utils/training.py

```diff
@@ -34,8 +36,48 @@ def compose_datasets(datasets, tf_dataset_cfg):
         train_data = train_data.shuffle(tf_dataset_cfg["shuffle_buffer_size"])
     if tf_dataset_cfg["cache"]:
         train_data = train_data.cache()
-    train_data = train_data.batch(tf_dataset_cfg["train_batch_size"])
-    val_data = val_data.batch(tf_dataset_cfg["val_batch_size"])
+
+    if tf_dataset_cfg['smart_batching_step'] is None:
+        train_data = train_data.batch(tf_dataset_cfg["train_batch_size"])
+        val_data = val_data.batch(tf_dataset_cfg["val_batch_size"])
+    else:
+        def element_to_bucket_id(*args):
+            seq_length = element_length_func(*args)
+
+            boundaries = list(bucket_boundaries)
+            buckets_min = [np.iinfo(np.int32).min] + boundaries
+            buckets_max = boundaries + [np.iinfo(np.int32).max]
+            conditions_c = math_ops.logical_and(
+            math_ops.less_equal(buckets_min, seq_length),
+            math_ops.less(seq_length, buckets_max))
+            bucket_id = math_ops.reduce_min(array_ops.where(conditions_c))
+
+            return bucket_id
+
+        def reduce_func(unused_arg, dataset):
```

Comment:
Since `reduce_func()` now is assumed to accept `batch_size` as a last argument, please update it in its definition  
```suggestion
        def reduce_func(unused_arg, dataset, batch_size):
```